packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,169 @@
+import { XhrPerformanceResourceTiming } from './zone-types';
+
+/**
+ * Get Browser's performance resource timing data associated to a XHR.
+ * Some XHRs might have two or one performance resource timings as one of them
+ * is the CORS pre-flight request and the second is related to the actual HTTP
+ * request.
+ * In overall, the algorithm to get this data takes the best fit for the span,
+ * this means the closest performance resource timing to the span start/end
+ * performance times is the returned value.
+ * @param xhrUrl
+ * @param span
+ */
+export function getXhrPerfomanceData(
+  xhrUrl: string,
+  span: Span
+): XhrPerformanceResourceTiming | undefined {
+  const filteredPerfEntries = getPerfResourceEntries(xhrUrl, span);
+  const possibleEntries = getPossiblePerfResourceEntries(filteredPerfEntries);
+  const bestEntry = getBestPerfResourceTiming(possibleEntries, span);
+  return bestEntry;
+}
+
+/**
+ * First step for the algorithm. Filter the Performance Resource Timings by the
+ * name (it should match the XHR URL), additionally, the start/end timings of
+ * every performance entry should fit within the span start/end timings.
+ * These filtered performance resource entries are considered as possible
+ * entries associated to the xhr.
+ * Those are possible because there might be more than two entries that pass the
+ * filter.
+ * @param xhrUrl
+ * @param span
+ */
+export function getPerfResourceEntries(
+  xhrUrl: string,
+  span: Span
+): PerformanceResourceTiming[] {
+  return performance
+    .getEntriesByType('resource')
+    .filter(entry =>
+      isPerfEntryPartOfXhr(entry as PerformanceResourceTiming, xhrUrl, span)
+    ) as PerformanceResourceTiming[];
+}
+
+/**
+ * Second step for the 'Performance resource timings selector algorithm'.
+ * As the XHR could cause a CORS pre-flight request, we have to look for
+ * possible performance entries either containing cors preflight timings or not.
+ * A possible entry with cors data is when a resource timing entry does not
+ * overlap timings with other resource timing entry. Also, a possible entry
+ * without cors resource timing is when that resource timing entry is not
+ * 'paired' with any other entry.
+ * Thus, for this step traverse the array of resource entries and for every
+ * entry check if it is a possible performance resource entry.
+ * @param perfEntries
+ */
+export function getPossiblePerfResourceEntries(
+  perfEntries: PerformanceResourceTiming[]
+): XhrPerformanceResourceTiming[] {
+  const possiblePerfEntries = new Array<XhrPerformanceResourceTiming>();
+  const pairedEntries = new Set<PerformanceResourceTiming>();
+  let entryI: PerformanceResourceTiming;
+  let entryJ: PerformanceResourceTiming;
+  // As this part of the algorithm traverses the array twice, although,
+  // this array is not large as the performance resource entries buffer is
+  // cleared when there are no more running XHRs.
+  for (let i = 0; i < perfEntries.length; i++) {
```

Comment:
Can you use `for (const .. of` here?

Reply:
It is not possible as we need to look the possible pair for the entryI comparing with the entries in front of it. This is possible as the entries array is already sorted with the `startTime`. Added some comments.